### PR TITLE
switch the mongo driver from mgo to official mongo go driver

### PR DESCRIPTION
**Pull request by 3pCode**
To support the 3.6 connection string (and Atlas), switch the mongo store to use the official mongo go driver.

Testing performed:
Modified cayley to pick up this branch of hidalgo
Run unit tests
Run mongo unit tests:
go test -v -tags docker -timeout 2h -run=TestNoSQL/mongo ./...

### CODE REVIEW

File: legacy/nosql/mongo/mongo.go [hunk 1]
```diff
@@ -43,83 +47,119 @@ func init() {
...
+	c := db.db.Collection(col)
 	compPK := len(primary.Fields) != 1
 	if compPK {
-		err := c.EnsureIndex(mgo.Index{
-			Key:    []string(primary.Fields),
-			Unique: true,
-		})
+		indexView := c.Indexes()
+		indexOptions := options.Index().SetUnique(true)
+		keys := bson.D{}
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
		keys := make(bson.D, 0, len(primary.Fields))
```

File: legacy/nosql/mongo/mongo.go [hunk 2]
```diff
@@ -43,83 +47,119 @@ func init() {
...
-			Key:        []string(ind.Fields),
-			Unique:     false,
-			Background: true,
-			Sparse:     true,
-		})
+		indexView := c.Indexes()
+
+		indexOptions := options.Index().SetUnique(false).SetSparse(true).SetBackground(true)
+
+		keys := bson.D{}
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
		keys := make(bson.D, 0, len(ind.Fields))
```

File: legacy/nosql/mongo/mongo.go [hunk 3]
```diff
@@ -43,83 +47,119 @@ func init() {
...
-			Unique: true,
-		})
+		indexView := c.Indexes()
+		indexOptions := options.Index().SetUnique(true)
+		keys := bson.D{}
+
+		for _, field := range primary.Fields {
+			keys = append(keys, primitive.E{Key: field, Value: bsonx.Int32(1)})
+		}
+		index := mongo.IndexModel{}
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
		index := mongo.IndexModel{
			Keys: keys,
			Options: indexOptions,
		}
```
Same below.

File: legacy/nosql/mongo/mongo.go [hunk 4]
```diff
@@ -170,20 +210,25 @@ func fromBsonValue(v interface{}) nosql.Value {
...
+	case int32:
+		return nosql.Int(v)
 	case float64:
 		return nosql.Float(v)
 	case bool:
 		return nosql.Bool(v)
 	case time.Time:
 		return nosql.Time(v)
+	case primitive.DateTime:
+		return nosql.Time(time.Unix(int64(v)/1000, int64(v)%1000*1000000))
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
		const ms = int64(time.Millisecond)
		return nosql.Time(time.Unix(int64(v)/1000, int64(v)%1000*ms))
```

File: legacy/nosql/mongo/mongo.go [hunk 5]
```diff
@@ -297,21 +342,25 @@ func (db *DB) Insert(ctx context.Context, col string, key nosql.Key, d nosql.Doc
...
 func (db *DB) FindByKey(ctx context.Context, col string, key nosql.Key) (nosql.Document, error) {
 	c := db.colls[col]
-	var m bson.M
-	err := c.c.FindId(compKey(key)).One(&m)
-	if err == mgo.ErrNotFound {
+
+	res := c.c.FindOne(ctx, bson.M{"_id": compKey(key)})
+	m := &bson.M{}
+
+	err := res.Decode(m)
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
	var m bson.M
	err := res.Decode(&m)
```
Same below.

File: legacy/nosql/mongo/mongo.go [hunk 6]
```diff
@@ -396,51 +445,104 @@ func (q *Query) Limit(n int) nosql.Query {
...
+	var m = &bson.M{}
+	cursor, err := q.build()
+
+	if err != nil {
 		return nil, err
 	}
-	return q.c.convDoc(m), nil
+	defer cursor.Close(ctx)
+
+	if cursor.Next(ctx) {
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
	if !cursor.Next(ctx) {
		return nil, nosql.ErrNotFound
	}
	// the rest of the code
```

File: legacy/nosql/mongo/mongo.go [hunk 7]
```diff
@@ -396,51 +445,104 @@ func (q *Query) Limit(n int) nosql.Query {
...
+	it  *mongo.Cursor
 	res bson.M
 }
 
 func (it *Iterator) Next(ctx context.Context) bool {
-	it.res = make(bson.M)
-	return it.it.Next(&it.res)
+	elem := make(bson.M)
+
+	if it.it != nil {
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
	if it.it != nil {
		return false
	}
	elem := make(bson.M)
	// the rest of the code
```

File: legacy/nosql/mongo/mongo.go [hunk 8]
```diff
@@ -396,51 +445,104 @@ func (q *Query) Limit(n int) nosql.Query {
...
+		return next
+	} else {
+		return false
+	}
 }
 func (it *Iterator) Err() error {
-	return it.it.Err()
+	if it.it != nil {
+		return it.it.Err()
+	} else {
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]: `else` can be dropped

File: legacy/nosql/mongo/mongo.go [hunk 9]
```diff
@@ -527,15 +630,17 @@ func (u *Update) Upsert(d nosql.Document) nosql.Update {
...
 	if u.upsert != nil {
 		if len(u.upsert) != 0 {
+			updateOptions.SetUpsert(true)
 			u.update["$setOnInsert"] = u.upsert
 		}
-		_, err = u.col.c.UpsertId(key, u.update)
+		_, err = u.col.c.UpdateOne(ctx, idFilter, u.update, updateOptions)
 	} else {
-		err = u.col.c.UpdateId(key, u.update)
+		_, err = u.col.c.UpdateOne(ctx, idFilter, u.update, updateOptions)
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]: It's the same as the `UpdateOne` above, right? So we can remove the duplication an move it out of the `if` statement.

File: legacy/nosql/mongo/test/mongotest.go [hunk 10]
```diff
@@ -33,21 +36,29 @@ func MongoVersion(vers string) nosqltest.Database {
...
 
-			addr := cont.GetHostPort("27017/tcp")
-
+			addr := fmt.Sprintf("mongodb://%s", cont.GetHostPort("27017/tcp"))
 			err = pool.Retry(func() error {
-				sess, err := mgo.Dial(addr)
+				sess, err := gomongo.NewClient(options.Client().ApplyURI(addr))
+
+				if err != nil {
+					t.Fatal(err)
```
Comment (dennwc) [inline legacy/nosql/mongo/test/mongotest.go]:
```suggestion
					return err
```

File: legacy/nosql/mongo/test/mongotest.go [hunk 11]
```diff
@@ -33,21 +36,29 @@ func MongoVersion(vers string) nosqltest.Database {
...
+				}
+
+				if err == nil {
```
Comment (dennwc) [inline legacy/nosql/mongo/test/mongotest.go]: `if` can be dropped, assuming suggestion above

File: legacy/nosql/mongo/mongo.go [hunk 12]
```diff
@@ -527,16 +628,18 @@ func (u *Update) Upsert(d nosql.Document) nosql.Update {
 	return u
 }
 func (u *Update) Do(ctx context.Context) error {
-	key := compKey(u.key)
+	idFilter := bson.M{idField: compKey(u.key)}
 	var err error
+	updateOptions := options.Update()
 	if u.upsert != nil {
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
	if u.upsert != nil && len(u.upsert) != 0 {
```

File: legacy/nosql/mongo/mongo.go [hunk 13]
```diff
@@ -527,16 +628,18 @@ func (u *Update) Upsert(d nosql.Document) nosql.Update {
...
+			updateOptions.SetUpsert(true)
 			u.update["$setOnInsert"] = u.upsert
 		}
-		_, err = u.col.c.UpsertId(key, u.update)
-	} else {
-		err = u.col.c.UpdateId(key, u.update)
-	}
+		
+	} 
+	_, err = u.col.c.UpdateOne(ctx, idFilter, u.update, updateOptions)
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
	_, err := u.col.c.UpdateOne(ctx, idFilter, u.update, updateOptions)
```

File: legacy/nosql/mongo/test/mongotest.go [hunk 14]
```diff
@@ -33,21 +36,27 @@ func MongoVersion(vers string) nosqltest.Database {
...
 
-			addr := cont.GetHostPort("27017/tcp")
-
+			addr := fmt.Sprintf("mongodb://%s", cont.GetHostPort("27017/tcp"))
 			err = pool.Retry(func() error {
-				sess, err := mgo.Dial(addr)
+				sess, err := gomongo.NewClient(options.Client().ApplyURI(addr))
+
+				if err != nil {
+					return err;
```
Comment (dennwc) [inline legacy/nosql/mongo/test/mongotest.go]:
```suggestion
					return err
```

File: legacy/nosql/mongo/mongo.go [hunk 15]
```diff
@@ -396,51 +444,103 @@ func (q *Query) Limit(n int) nosql.Query {
...
 		return nil, err
 	}
-	return q.c.convDoc(m), nil
+	defer cursor.Close(ctx)
+
+	if !cursor.Next(ctx) {
+		return nil, nosql.ErrNotFound
+	}
+	if err := cursor.Decode(m); err != nil {
+			return nil, err
```
Comment (dennwc) [inline legacy/nosql/mongo/mongo.go]:
```suggestion
		return nil, err
```
`gofmt`?

File: legacy/nosql/mongo/mongo.go (hunk 1 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated.

File: legacy/nosql/mongo/mongo.go (hunk 2 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 3 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 4 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 5 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 6 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 7 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated, this one I think your suggested change you meant it.it == nil { return false}

File: legacy/nosql/mongo/mongo.go (hunk 8 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated.

File: legacy/nosql/mongo/mongo.go (hunk 9 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: yup, updated.

File: legacy/nosql/mongo/test/mongotest.go (hunk 10 above)
Comment (adamst-q) [reply legacy/nosql/mongo/test/mongotest.go]: updated

File: legacy/nosql/mongo/test/mongotest.go (hunk 11 above)
Comment (adamst-q) [reply legacy/nosql/mongo/test/mongotest.go]: updated.

**Comment by adamst-q** [comment]
thanks Den updated, if you could do one more pass over.

File: legacy/nosql/mongo/mongo.go [hunk 16]
```diff
@@ -396,51 +444,103 @@ func (q *Query) Limit(n int) nosql.Query {
...
+	if err == nil {
+		it.res = elem
+	}
+
+	return next
+	
 }
 func (it *Iterator) Err() error {
-	return it.it.Err()
+	if it.it != nil {
```
[the remark this answers is not in the document]
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated.

File: legacy/nosql/mongo/mongo.go [hunk 17]
```diff
@@ -396,51 +444,103 @@ func (q *Query) Limit(n int) nosql.Query {
...
 	res bson.M
 }
 
 func (it *Iterator) Next(ctx context.Context) bool {
-	it.res = make(bson.M)
-	return it.it.Next(&it.res)
+	elem := make(bson.M)
+
+	if it.it == nil {
+		return false;
```
[the remark this answers is not in the document]
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go [hunk 18]
```diff
@@ -396,51 +444,103 @@ func (q *Query) Limit(n int) nosql.Query {
...
+	}
+	next := it.it.Next(ctx)
```
[the remark this answers is not in the document]
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 12 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 13 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated

File: legacy/nosql/mongo/test/mongotest.go (hunk 14 above)
Comment (adamst-q) [reply legacy/nosql/mongo/test/mongotest.go]: updated

File: legacy/nosql/mongo/mongo.go (hunk 15 above)
Comment (adamst-q) [reply legacy/nosql/mongo/mongo.go]: updated, formated.

**Pull request merged**